### PR TITLE
Various improvements around dependency resolution; better logging

### PR DESCRIPTION
**Pull request by anutron**
- Now fully supports YAML dependency graphs
- Supports components referencing other components in the same package 
  with the `/SomethingInThisPackage` or just `SomethingInThisPackage` 
  convention
- Adds additional `--verbose` logging 
- Adds support for `--show-packages` flag which will list all discovered packages and their components
- Adds support for configurations that use `only: ['PackageName/*']` wildcards to include entire packages
- Improves logging when a package is unable to be resolved in the dependency graph
- Added documentation for various flags and options

### CODE REVIEW

File: tasks/packager.js [hunk 1]
```diff
@@ -15,11 +15,14 @@ module.exports = function(grunt) {
 	var DESC_REGEXP = /\/\*\s*^---([.\s\S]*)^(?:\.\.\.|---)\s*\*\//m;
 	var SL_STRIP_EXP = ['\\/[\\/*]\\s*<', '>(.*?)<\\/', '>(?:\\s*\\*\\/)?'];
 	var ML_STRIP_EXP = ['\\/[\\/*]\\s*<', '>([.\\s\\S]*?)<\\/', '>(?:\\s*\\*\\/)?'];
+	var PACKAGE_DOT_STAR = /(.*)\/\*$/;
 
-	function validDefinition(object) {
-		return 'name' in object && 'provides' in object;
+	// ensures the definition has a name and a provision
```
Comment (ibolmo) [inline tasks/packager.js:20]: the method name would suffice. same for the other comments you've added.

Comment (anutron) [inline tasks/packager.js:20]: As someone who just worked through this code, a few more comments would really have been welcome. I realize these are perhaps not significantly value-add, but really, this kind of thick-business-logic code used to graph an esoteric dependency system is always unpleasant to untangle (speaking as someone who has written it thrice).

I never could figure out with the moo group hates comments so much. They make code better.

File: tasks/packager.js [hunk 2]
```diff
@@ -31,58 +34,97 @@ module.exports = function(grunt) {
...
-		var options = this.options({
-			only: false,
-			strip: [],
-			separator: grunt.util.linefeed
-		});
+		// loads a component and its dependencies
+		// if the key given is a package and a wildcard, loads all of them
+		// e.g. `Package/Component` OR `Package/*`
+		var loadComponent = function(key){
+			var wildCardMatch = key.match(PACKAGE_DOT_STAR)
```
Comment (ibolmo) [inline tasks/packager.js]: missing ';'

File: tasks/packager.js [hunk 3]
```diff
@@ -31,58 +34,97 @@ module.exports = function(grunt) {
...
+			if (wildCardMatch){
+				packages[wildCardMatch[1]].forEach(loadComponent);
+			} else {
```
Comment (ibolmo) [inline tasks/packager.js:97]:
``` js
else if (key in registry...) resolveDeps....
else throw new ...
```

File: tasks/packager.js [hunk 4]
```diff
@@ -104,12 +145,29 @@ module.exports = function(grunt) {
...
-			(options.only ? toArray(options.only) : set).forEach(function(key){
-				if (!(key in registry)) throw new Error('Missing key: ' + key);
-				resolveDeps(registry[key]);
+			set.forEach(function(key){
+				var definition = registry[key];
+				if (!packages[definition.package]) packages[definition.package] = []
+				packages[definition.package].push(key);
 			});
 
+			if (grunt.option('show-packages')){
```
Comment (ibolmo) [inline tasks/packager.js]: I'd say drop show-packages, and roll this into the set.forEach. If it'll be verbose, let's just be verbose. ;)

File: tasks/packager.js [hunk 5]
```diff
@@ -104,12 +145,29 @@ module.exports = function(grunt) {
...
+				grunt.log.verbose.writeln('Loaded packages:')
+				for (var p in packages){
+					grunt.log.verbose.writeln('Package: ', p);
+					grunt.log.verbose.writeln(' ', packages[p])
+				}
+			}
+			// support global options.only as well as .only per build
+			var only = options.only || f.only;
+
+			if (only) grunt.log.verbose.writeln('compiling', f.dest, 'with dependencies:', only);
```
Comment (ibolmo) [inline tasks/packager.js]: Go ahead and simplify into `'compiling' , f.dest, 'with deps:', only || 'all'`

File: tasks/packager.js [hunk 6]
```diff
@@ -119,6 +177,9 @@ module.exports = function(grunt) {
 					.replace(RegExp(ML_STRIP_EXP.join(block), 'gm'), '');
 			});
 
+			if (only) grunt.log.verbose.ok('successfully compiled', f.dest, 'with dependencies:', only);
+			else grunt.log.verbose.ok('successfully compiled', f.dest);
```
Comment (ibolmo) [inline tasks/packager.js]: Go ahead and simplify into `'compiled' , f.dest, 'with deps:', only || 'all'`

**Comment by ibolmo** [comment]
Could you add specs for the YAML header extensions?

e.g.:
1. `/Component`
2. `Package/*`

**Comment by anutron** [comment]
I'll address comments and re-push. I'm neck deep in trying to get another project up on github (that uses this packager) so don't know if I'll get to the specs immediately. You can hold off on pulling it until I add it or whatever you want.

**Comment by ibolmo** [comment]
Yeah, I'll wait. Not that it's not good code. Looks solid. Just want to
make sure it's BC and that any further changes will be BC with the new
extensions

**Comment by anutron** [comment]
Pushed all your suggested code changes. Specs to follow eventually.

@SergioCrisostomo if you wanted to help out with that it'd rock.

**Comment by SergioCrisostomo** [comment]
Nice work @anutron !
I think the packager is now ready to serve the builder of the future node website, nice!

I am happy to contribute with specs. 

Made 2 commits, welcome to include them if you guys aprove them: https://github.com/SergioCrisostomo/grunt-packager/commits/anutron

**Comment by ibolmo** [comment]
if specs are green on your end Sergio, go ahead and send PR. I can just
pull yours since yours has aaron's changes.

On Sat, Jun 28, 2014 at 8:43 AM, Sergio Crisostomo <notifications@github.com

**Comment by ibolmo** [comment]
Go ahead and bump the version too. Make it a minor up.

**Comment by SergioCrisostomo** [comment]
Yes Sir!
`0.3.0` on the way...

**Comment by ibolmo** [comment]
See #6 which is #5 + Sergio.

**Comment by anutron** [comment]
Yay!

**Comment by SergioCrisostomo** [comment]
:+1:

**Pull request merged**